src/patterns.rs

```diff
@@ -20,187 +22,151 @@ use ffi::{
-        })
+    pub fn get_type(&self) -> PatternType {
+        unsafe {
+            ffi::cairo_pattern_get_type(self.pointer).into()
+        }
     }
-}
-
-pub trait PatternTrait {
-    type PatternType;
 
-    fn as_ptr(&self) -> *mut cairo_pattern_t;
-
-    fn ensure_status(&self) {
+    pub fn ensure_status(&self) {
         self.status().ensure_valid();
     }
 
-    fn status(&self) -> Status {
+    pub fn status(&self) -> Status {
         unsafe {
-            Status::from(ffi::cairo_pattern_status(self.as_ptr()))
+            Status::from(ffi::cairo_pattern_status(self.pointer))
         }
     }
 
-    fn get_reference_count(&self) -> isize {
+    pub fn get_reference_count(&self) -> isize {
         unsafe {
-            ffi::cairo_pattern_get_reference_count(self.as_ptr()) as isize
+            ffi::cairo_pattern_get_reference_count(self.pointer) as isize
         }
     }
 
-    fn set_extend(&self, extend: Extend) {
+    pub fn set_extend(&self, extend: Extend) {
         unsafe {
-            ffi::cairo_pattern_set_extend(self.as_ptr(), extend.into())
+            ffi::cairo_pattern_set_extend(self.pointer, extend.into())
         }
     }
 
-    fn get_extend(&self) -> Extend {
+    pub fn get_extend(&self) -> Extend {
         unsafe {
-            Extend::from(ffi::cairo_pattern_get_extend(self.as_ptr()))
+            Extend::from(ffi::cairo_pattern_get_extend(self.pointer))
         }
     }
 
-    fn set_filter(&self, filter: Filter) {
+    pub fn set_filter(&self, filter: Filter) {
         unsafe {
-            ffi::cairo_pattern_set_filter(self.as_ptr(), filter.into())
+            ffi::cairo_pattern_set_filter(self.pointer, filter.into())
         }
     }
 
-    fn get_filter(&self) -> Filter {
+    pub fn get_filter(&self) -> Filter {
         unsafe {
-            Filter::from(ffi::cairo_pattern_get_filter(self.as_ptr()))
+            Filter::from(ffi::cairo_pattern_get_filter(self.pointer))
         }
     }
 
-    fn set_matrix(&self, matrix: Matrix) {
+    pub fn set_matrix(&self, matrix: Matrix) {
         unsafe {
-            ffi::cairo_pattern_set_matrix (self.as_ptr(), &matrix)
+            ffi::cairo_pattern_set_matrix (self.pointer, matrix.ptr())
         }
     }
 
-    fn get_matrix(&self) -> Matrix {
+    pub fn get_matrix(&self) -> Matrix {
         let mut matrix = Matrix::null();
         unsafe {
-            ffi::cairo_pattern_get_matrix(self.as_ptr(), &mut matrix);
+            ffi::cairo_pattern_get_matrix(self.pointer, matrix.mut_ptr());
         }
         matrix
     }
+}
 
-    unsafe fn from_raw_full(pointer: *mut cairo_pattern_t) -> Self::PatternType;
-
-    unsafe fn from_raw_none(pointer: *mut cairo_pattern_t) -> Self::PatternType {
-        ffi::cairo_pattern_reference(pointer);
-        Self::from_raw_full(pointer)
+impl Clone for Pattern {
+    fn clone(&self) -> Self {
+        Pattern {
+            pointer: unsafe {
+                ffi::cairo_pattern_reference(self.pointer)
```

Comment:
Shouldn't you check if the returned pointer is not NULL in here?

Reply:
This function returns the same pointer that it received, which must be non-NULL. https://www.cairographics.org/manual/cairo-cairo-pattern-t.html#cairo-pattern-reference